README.md

```diff
@@ -1,28 +1,182 @@
 
 ```ruby
 gem "velveteen"
 ```
 
-and run `bundle install` from your shell.
+Run the bundle command to install it.
+
+Define a worker:
+
+```ruby
+# do_something.rb
+require "velveteen"
+
+Velveteen::Config.exchange_name = "velveteen_development"
+
+class DoSomething < Velveteen::Worker
+  self.routing_key = "something.do"
+
+  def perform
+    # do something
+  end
+end
+```
+
+Run it:
+
+```shell
+velveteen work do_something.rb DoSomething
+```
+
+Within the worker, you have access to the following:
+
+* `message` - A representation of the Rabbit message, which responds to:
+  * `body` - the raw message body
+  * `data` - the parsed JSON representation of the body
+  * `delivery_info` - the delivery info from Rabbit
+  * `headers` - the headers from Rabbit, from `properties.headers`
+  * `properties` - the properties Rabbit
+* `publish(payload, [options])` - Publishes the message to the exchange
+
+  The headers from the message the worker is consuming will be automatically
+  included when publishing new messages.
+
+## Pipeline design
+
+In general, try to follow the best practices laid out in thoughtbot's [data
+guide].
+
+[data guide]: https://github.com/thoughtbot/guides/tree/master/data
+
+When using Velveteen in particular, try to follow:
+
+* Keep workers focused on a single task and interact with at most one service
+  (e.g. S3, Postgres, third-party API).
+* Name workers after the actions they perform. E.g. a worker that fetches a
+  user's most recent GitHub commits could be called `FetchUserCommits`.
+  Velveteen will use the same name for the worker's queue.
+* In general, routing keys should be past tense events. E.g. `FetchUserCommits`
+  could publish the API response with a routing key of `user.commits.fetched`.
+  This documents what already happened and makes no assumptions about what
+  happens next. If a worker is the first step in a pipeline, e.g. runs a
+  schedule, use present tense - `user.commits.fetch`.
+
+## Message validation
+
+Velveteen can validate incoming messages with a JSON Schema.
+
+```ruby
+Velveteen::Config.schema_directory = "app/schemas"
+
+class FetchUserCommits < Velveteen::Worker
+  ...
+  self.message_schema = "fetch_user_commits.json"
+
+  def perform
+    # do something with the GitHub API
+  end
+end
+```
+
+_app/schemas/fetch_user_commits.json_
+```json
+{
+  "type": "object",
+  "required": ["username"],
+  "properties": {
+    "username": {"type": "string"}
+  }
+}
+```
+
+When a message is added to this worker's queue and it does not match the schema,
+a `Velveteen::InvalidMessage` exception will be raised and handled by the
+configured error handler. The worker's `perform` method will not be invoked.
+
+## Rate limiting
+
+Limiting workers to processing a certain number of messages per minute is
+supported. This is implemented by periodically publishing tokens to a dedicated
+queue (with a max size of 1), where the worker must successfully take a token
+before it begins.
```

Comment:
This comment has no bearing on the README, but this is a cool strategy. I'm curious how it came about.